# Lending Club Case Study

## Import libraries

Importing the necessary libraries for performing Exploratory Data Analysis

In [7]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Data Visualization 

With any data, the first step would be to visualize the data and understand the content. Try to get a gist of the fields present, data present, cleansing requirements and so on.

## Uploading the data

Upload the data into the pandas dataframe for performing all analysis

In [39]:
data_set = pd.read_csv("loan.csv")
data_set.head(10)

/var/folders/cq/3_v0kxk56d79810f6w5gbnkr0000gn/T/ipykernel_81915/2955158840.py:1: DtypeWarning: Columns (47) have mixed types. Specify dtype option on import or set low_memory=False.
  data_set = pd.read_csv("loan.csv")


id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599       5000         5000           4975.0   36 months   
1  1077430    1314167       2500         2500           2500.0   60 months   
2  1077175    1313524       2400         2400           2400.0   36 months   
3  1076863    1277178      10000        10000          10000.0   36 months   
4  1075358    1311748       3000         3000           3000.0   60 months   
5  1075269    1311441       5000         5000           5000.0   36 months   
6  1069639    1304742       7000         7000           7000.0   60 months   
7  1072053    1288686       3000         3000           3000.0   36 months   
8  1071795    1306957       5600         5600           5600.0   60 months   
9  1071570    1306721       5375         5375           5350.0   60 months   

  int_rate  installment grade sub_grade  ... num_tl_90g_dpd_24m  \
0   10.65%       162.87     B        B2  ...                NaN   
1   15.27%        59.83     C        C4  ...                NaN   
2   15.96%        84.33     C        C5  ...                NaN   
3   13.49%       339.31     C        C1  ...                NaN   
4   12.69%        67.79     B        B5  ...                NaN   
5    7.90%       156.46     A        A4  ...                NaN   
6   15.96%       170.08     C        C5  ...                NaN   
7   18.64%       109.43     E        E1  ...                NaN   
8   21.28%       152.39     F        F2  ...                NaN   
9   12.69%       121.45     B        B5  ...                NaN   

  num_tl_op_past_12m pct_tl_nvr_dlq  percent_bc_gt_75 pub_rec_bankruptcies  \
0                NaN            NaN               NaN                  0.0   
1                NaN            NaN               NaN                  0.0   
2                NaN            NaN               NaN                  0.0   
3                NaN            NaN               NaN                  0.0   
4                NaN            NaN               NaN                  0.0   
5                NaN            NaN               NaN                  0.0   
6                NaN            NaN               NaN                  0.0   
7                NaN            NaN               NaN                  0.0   
8                NaN            NaN               NaN                  0.0   
9                NaN            NaN               NaN                  0.0   

  tax_liens tot_hi_cred_lim total_bal_ex_mort total_bc_limit  \
0       0.0             NaN               NaN            NaN   
1       0.0             NaN               NaN            NaN   
2       0.0             NaN               NaN            NaN   
3       0.0             NaN               NaN            NaN   
4       0.0             NaN               NaN            NaN   
5       0.0             NaN               NaN            NaN   
6       0.0             NaN               NaN            NaN   
7       0.0             NaN               NaN            NaN   
8       0.0             NaN               NaN            NaN   
9       0.0             NaN               NaN            NaN   

  total_il_high_credit_limit  
0                        NaN  
1                        NaN  
2                        NaN  
3                        NaN  
4                        NaN  
5                        NaN  
6                        NaN  
7                        NaN  
8                        NaN  
9                        NaN  

[10 rows x 111 columns]

## Describe the data 

Use the commands like describe, shape command to understand the field analytics

In [40]:
data_set.describe()

id     member_id     loan_amnt   funded_amnt  \
count  3.971700e+04  3.971700e+04  39717.000000  39717.000000   
mean   6.831319e+05  8.504636e+05  11219.443815  10947.713196   
std    2.106941e+05  2.656783e+05   7456.670694   7187.238670   
min    5.473400e+04  7.069900e+04    500.000000    500.000000   
25%    5.162210e+05  6.667800e+05   5500.000000   5400.000000   
50%    6.656650e+05  8.508120e+05  10000.000000   9600.000000   
75%    8.377550e+05  1.047339e+06  15000.000000  15000.000000   
max    1.077501e+06  1.314167e+06  35000.000000  35000.000000   

       funded_amnt_inv   installment    annual_inc           dti  \
count     39717.000000  39717.000000  3.971700e+04  39717.000000   
mean      10397.448868    324.561922  6.896893e+04     13.315130   
std        7128.450439    208.874874  6.379377e+04      6.678594   
min           0.000000     15.690000  4.000000e+03      0.000000   
25%        5000.000000    167.020000  4.040400e+04      8.170000   
50%        8975.000000    280.220000  5.900000e+04     13.400000   
75%       14400.000000    430.780000  8.230000e+04     18.600000   
max       35000.000000   1305.190000  6.000000e+06     29.990000   

        delinq_2yrs  inq_last_6mths  ...  num_tl_90g_dpd_24m  \
count  39717.000000    39717.000000  ...                 0.0   
mean       0.146512        0.869200  ...                 NaN   
std        0.491812        1.070219  ...                 NaN   
min        0.000000        0.000000  ...                 NaN   
25%        0.000000        0.000000  ...                 NaN   
50%        0.000000        1.000000  ...                 NaN   
75%        0.000000        1.000000  ...                 NaN   
max       11.000000        8.000000  ...                 NaN   

       num_tl_op_past_12m  pct_tl_nvr_dlq  percent_bc_gt_75  \
count                 0.0             0.0               0.0   
mean                  NaN             NaN               NaN   
std                   NaN             NaN               NaN   
min                   NaN             NaN               NaN   
25%                   NaN             NaN               NaN   
50%                   NaN             NaN               NaN   
75%                   NaN             NaN               NaN   
max                   NaN             NaN               NaN   

       pub_rec_bankruptcies  tax_liens  tot_hi_cred_lim  total_bal_ex_mort  \
count          39020.000000    39678.0              0.0                0.0   
mean               0.043260        0.0              NaN                NaN   
std                0.204324        0.0              NaN                NaN   
min                0.000000        0.0              NaN                NaN   
25%                0.000000        0.0              NaN                NaN   
50%                0.000000        0.0              NaN                NaN   
75%                0.000000        0.0              NaN                NaN   
max                2.000000        0.0              NaN                NaN   

       total_bc_limit  total_il_high_credit_limit  
count             0.0                         0.0  
mean              NaN                         NaN  
std               NaN                         NaN  
min               NaN                         NaN  
25%               NaN                         NaN  
50%               NaN                         NaN  
75%               NaN                         NaN  
max               NaN                         NaN  

[8 rows x 87 columns]

In [41]:
data_set.shape

(39717, 111)

## Categorize the data based on the consumption

Analyse the columns in the data set and categorise then based on the impact they have in our decision making process

In [42]:
data_set.columns

Index(['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv',
       'term', 'int_rate', 'installment', 'grade', 'sub_grade',
       ...
       'num_tl_90g_dpd_24m', 'num_tl_op_past_12m', 'pct_tl_nvr_dlq',
       'percent_bc_gt_75', 'pub_rec_bankruptcies', 'tax_liens',
       'tot_hi_cred_lim', 'total_bal_ex_mort', 'total_bc_limit',
       'total_il_high_credit_limit'],
      dtype='object', length=111)

### Fields present in the data set: 

#### Personal information 
- id
- member_id
- annual_inc
- annual_inc_joint
- emp_length
- emp_title
- grade
- home_ownership
- addr_state
- zip_code
- verification_status
- verified_status_joint

#### Loan Information

- title
- desc
- application_type
- avg_cur_bal
- acc_now_delinq
- acc_open_past_24mths
- chargeoff_within_12_mths
- collection_recovery_fee
- collections_12_mths_ex_med
- delinq_2yrs
- delinq_amnt
- fico_range_high
- fico_range_low
- funded_amnt
- funded_amnt_inv
- initial_list_status
- installment
- int_rate
- issue_d
- last_credit_pull_d
- last_fico_range_high
- last_fico_range_low
- last_pymnt_amnt
- last_pymnt_d
- loan_amnt
- loan_status
- mo_sin_old_il_acct
- mo_sin_old_rev_tl_op
- mo_sin_rcnt_rev_tl_op
- mo_sin_rcnt_tl
- mort_acc
- mths_since_last_delinq
- mths_since_last_major_derog
- mths_since_last_record
- mths_since_rcnt_il
- mths_since_recent_bc
- mths_since_recent_bc_dlq
- mths_since_recent_inq
- mths_since_recent_revol_delinq
- next_pymnt_d
- num_accts_ever_120_pd
- num_actv_bc_tl
- num_actv_rev_tl
- num_bc_sats
- num_bc_tl
- num_il_tl
- num_op_rev_tl
- num_rev_accts
- num_rev_tl_bal_gt_0
- num_sats
- num_tl_120dpd_2m
- num_tl_30dpd
- num_tl_90g_dpd_24m
- num_tl_op_past_12m
- open_acc
- open_acc_6m
- open_il_12m
- open_il_24m
- open_il_6m
- open_rv_12m
- open_rv_24m
- out_prncp
- out_prncp_inv
- pct_tl_nvr_dlq
- percent_bc_gt_75
- policy_code
- pub_rec
- pub_rec_bankruptcies
- purpose
- pymnt_plan
- recoveries
- revol_bal
- revol_util
- sub_grade
- tax_liens
- term
- tot_coll_amt
- tot_cur_bal
- tot_hi_cred_lim
- total_acc
- total_bal_ex_mort
- total_bal_il
- total_bc_limit
- total_cu_tl
- total_il_high_credit_limit
- total_pymnt
- total_pymnt_inv
- total_rec_int
- total_rec_late_fee
- total_rec_prncp
- total_rev_hi_lim  
- url


#### Credit Information
- dti
- dti_joint
- all_util
- bc_util
- il_util
- earliest_cr_line
- max_bal_bc
- bc_open_to_buy
- inq_fi
- inq_last_12m
- inq_last_6mths


## Data Cleansing 

#### Columns to be dropped 

It is clear that the data set has many columns that may not be useful in identifying risky applicants. Hence, we can remove them from the data set for easier analysis. 

The below data are relevant are personal and of no use for our analysis:

- url
- emp_title
- zip_code
- addr_state

The below Loan related columns are of no use to us in perfoming the analysis:
- last_credit_pull_d
- desc
- title
- out_prncp_inv 
- total_pymnt_inv 


The below information becomes only relevant after the loan approval process. Since we are not interested in already approved loans and the process around it, we can delete the below rows as well:

- delinq_2yrs
- earliest_cr_line
- inq_last_6mths
- acc_now_delinq
- open_acc
- pub_rec
- revol_bal
- revol_util
- total_acc
- out_prncp
- out_prncp_inv
- total_pymnt
- total_pymnt_inv
- total_rec_prncp
- total_rec_int
- total_rec_late_fee
- recoveries
- collection_recovery_fee
- last_pymnt_d
- last_pymnt_amnt
- next_pymnt_d
- last_credit_pull_d
- application_type
- pymnt_plan
- initial_list_status
- collections_12_mths_ex_med
- policy_code
- tax_liens

In [43]:
data_set.drop(["url","title","emp_title","zip_code","addr_state","last_credit_pull_d", "desc","out_prncp_inv","total_pymnt_inv","delinq_2yrs", "earliest_cr_line", "inq_last_6mths"
, "open_acc", "pub_rec","revol_bal", "revol_util", "total_acc", "out_prncp", "out_prncp_inv", "total_pymnt", "total_pymnt_inv"
, "total_rec_prncp","total_rec_int","total_rec_late_fee","recoveries","collection_recovery_fee","last_pymnt_d","last_pymnt_amnt"
               ,"next_pymnt_d","last_credit_pull_d","application_type","tax_liens", "pymnt_plan","policy_code", "acc_now_delinq", "initial_list_status", "collections_12_mths_ex_med" ], axis = 1, inplace = True)


In [44]:
data_set.shape

(39717, 77)

We reduced the columns from 111 to 77 by performing the above clean up. 

#### Null value columns removal 

There are many columns with null values only present in it. We can remove them from the data set, as they wont be of any help to us. 

In [45]:
data_set.isnull().sum()

id                                0
member_id                         0
loan_amnt                         0
funded_amnt                       0
funded_amnt_inv                   0
                              ...  
pub_rec_bankruptcies            697
tot_hi_cred_lim               39717
total_bal_ex_mort             39717
total_bc_limit                39717
total_il_high_credit_limit    39717
Length: 77, dtype: int64

In [46]:
data_set.dropna(axis = 1, how = 'all', inplace = True)
data_set.shape

(39717, 23)

We have now reduced the columns to 23 from then orignal number of 111. 

In [47]:
data_set.columns

Index(['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv',
       'term', 'int_rate', 'installment', 'grade', 'sub_grade', 'emp_length',
       'home_ownership', 'annual_inc', 'verification_status', 'issue_d',
       'loan_status', 'purpose', 'dti', 'mths_since_last_delinq',
       'mths_since_last_record', 'chargeoff_within_12_mths', 'delinq_amnt',
       'pub_rec_bankruptcies'],
      dtype='object')

### Removing Loans with not relevant status value 

Let us now remove the loans that are marked as CURRENT. Since the data set does not provide information on the EMI's missed we cannot use the data of CURRENT loans to perform our analysis. We will focus on FULLY PAID and CHARGED OFF Loans only 

In [55]:
data_set = data_set[data_set.loan_status != "Current"]
data_set.shape

(38577, 23)

In [56]:
data_set.loan_status.unique()

array(['Fully Paid', 'Charged Off'], dtype=object)

Let us now download the data set to analyse in excel and view the cleansed data

In [58]:
data_set.to_csv("loans_cleansed.csv")

### Removing columns with same value present in all rows

From the downloaded excel, it was clear that there are columns where all the rows have the same value. This will not help us in the analysis and hence can be deleted. 

In [61]:
data_set = data_set.drop(['delinq_amnt'], axis=1)

KeyError: "['delinq_amnt'] not found in axis"

In [62]:
data_set.shape

(38577, 22)